pkg/maps/nat/stats/stats.go

```diff
@@ -0,0 +1,291 @@
+	metrics natMetrics
+
+	db    *statedb.DB
+	table statedb.RWTable[NatMapStats]
+
+	logger   logrus.Entry
+	maxPorts int
+	config   Config
+	natMap4  nat.NatMap4
+	natMap6  nat.NatMap6
+	ipcache  *ipcache.IPCache
+}
+
+type NatMapStats struct {
+	Type       string
+	EgressIP   string
+	EndpointIP string
+	RemotePort uint16
+	Count      int
+	Nth        int
+}
+
+type params struct {
+	cell.In
+
+	Lifecycle cell.Lifecycle
+	DB        *statedb.DB
+	Table     statedb.RWTable[NatMapStats]
+	NatMap4   promise.Promise[nat.NatMap4]
+	NatMap6   promise.Promise[nat.NatMap6]
+	Metrics   natMetrics
+	Config    Config
+	Health    cell.Health
+}
+
+var (
+	kBucketsToStore = 32
+)
+
+func newStats(params params) (*Stats, error) {
+	if params.Config.NATMapStatInterval == 0 {
+		return nil, nil
+	}
+	maxAvailPorts := 65535 - (option.Config.NodePortMax + 1)
+	m := &Stats{
+		metrics:  params.Metrics,
+		config:   params.Config,
+		maxPorts: maxAvailPorts,
+		db:       params.DB,
+		table:    params.Table,
+	}
+	statsCtx, cancelStats := context.WithCancel(context.Background())
+	params.Lifecycle.Append(cell.Hook{
+		OnStart: func(hc cell.HookContext) error {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*120)
+			defer cancel()
+			nmap4, err := params.NatMap4.Await(ctx)
+			if err != nil {
+				if !errors.Is(err, nat.MapDisabled) {
+					return err
+				}
+			}
+			nmap6, err := params.NatMap6.Await(ctx)
+			if err != nil {
+				if !errors.Is(err, nat.MapDisabled) {
+					return err
+				}
+			}
+			m.natMap4 = nmap4
+			m.natMap6 = nmap6
+			if m.natMap4 == nil && m.natMap6 == nil {
+				return nil
+			}
+
+			m.run(statsCtx, params.Health.NewScope("nat-stats"))
+			return nil
+		},
+		OnStop: func(hc cell.HookContext) error {
+			cancelStats()
+			return nil
+		},
+	})
+	return m, nil
+}
+
+func upsertStat[KT tupleKey](m *Stats, topk *topk[KT], family nat.IPFamily) error {
+	tx := m.db.WriteTxn(m.table)
+	defer tx.Abort()
+
+	var errs error
+	iter, _ := m.table.All(tx)
+	for entry, _, ok := iter.Next(); ok; entry, _, ok = iter.Next() {
+		if entry.Type == family.String() {
+			_, _, err := m.table.Delete(tx, entry)
+			errors.Join(errs, err)
+		}
+	}
+
+	topk.popForEach(func(key KT, count, ith int) {
+		if ith == kBucketsToStore-1 {
+			m.metrics.updateLocalPorts(family, count, m.maxPorts)
+		}
+		extip, eip, proto, rport := key.tuple()
+		_, _, err := m.table.Insert(tx, NatMapStats{
+			Type:       family.String(),
+			EgressIP:   eip,
+			EndpointIP: extip,
+			RemotePort: rport,
+			Proto:      proto,
+			Count:      count,
+			Nth:        ith,
+		})
+		if err != nil {
+			errs = errors.Join(errs, err)
+		}
+	})
+	if errs != nil {
+		return fmt.Errorf("failures occured updating stats table, transaction will not be committed: %w", errs)
+	}
+	tx.Commit()
+	return nil
+}
+
+func (m *Stats) run(ctx context.Context, hr cell.Health) {
+	controller.NewManager().UpdateController("nat-stats", controller.ControllerParams{
```

Comment:
Any reason not to use the timer jobs? We seem to be preferring those over controllers, at the least for new code.

Reply:
This seems to still be here. Didn't push yet?